ethdb/kv_snapshot.go

```diff
@@ -0,0 +1,502 @@
+	if !ok {
+		return fmt.Errorf("%T doesn't implement ethdb.TxMigrator interface", s.dbTX)
+	}
+	return db.CreateBucket(bucket)
+}
+
+func (s *snapshotTX) ExistsBucket(bucket string) bool {
+	db, ok := s.dbTX.(BucketMigrator)
+	if !ok {
+		return false
+	}
+	return db.ExistsBucket(bucket)
+}
+
+func (s *snapshotTX) ClearBucket(bucket string) error {
+	db, ok := s.dbTX.(BucketMigrator)
+	if !ok {
+		return fmt.Errorf("%T doesn't implement ethdb.TxMigrator interface", s.dbTX)
+	}
+	return db.ClearBucket(bucket)
+}
+
+func (s *snapshotTX) ExistingBuckets() ([]string, error) {
+	db, ok := s.dbTX.(BucketMigrator)
+	if !ok {
+		return []string{}, fmt.Errorf("%T doesn't implement ethdb.TxMigrator interface", s.dbTX)
+	}
+	return db.ExistingBuckets()
+}
+
+func NewSnapshotKV() snapshotOpts {
+	return snapshotOpts{
+		forBuckets: make(map[string]dbutils.BucketConfigItem),
+	}
+}
+
+type SnapshotKV struct {
+	db         KV
+	snapshotDB KV
+	forBuckets map[string]dbutils.BucketConfigItem
+}
+
+type snapshotOpts struct {
+	db         KV
+	snapshot   KV
+	forBuckets map[string]dbutils.BucketConfigItem
+}
+
+func (opts snapshotOpts) SnapshotDB(db KV) snapshotOpts {
+	opts.snapshot = db
+	return opts
+}
+
+func (opts snapshotOpts) DB(db KV) snapshotOpts {
+	opts.db = db
+	return opts
+}
+
+func (opts snapshotOpts) For(bucket string, config dbutils.BucketConfigItem) snapshotOpts {
+	opts.forBuckets[bucket] = config
+	return opts
+}
+
+func (opts snapshotOpts) MustOpen() KV {
+	return &SnapshotKV{
+		snapshotDB: opts.snapshot,
+		db:         opts.db,
+		forBuckets: opts.forBuckets,
+	}
+}
+
+func (s *SnapshotKV) View(ctx context.Context, f func(tx Tx) error) error {
+	dbTx, err := s.db.Begin(ctx, nil, false)
+	if err != nil {
+		return err
+	}
+
+	t := &snapshotTX{
+		dbTX: dbTx,
+		snTX: newVirtualTx(func() (Tx, error) {
+			return s.snapshotDB.Begin(ctx, nil, false)
+		}, s.forBuckets),
+		forBuckets: s.forBuckets,
+	}
+	defer t.Rollback()
+	return f(t)
+}
+
+func (s *SnapshotKV) Update(ctx context.Context, f func(tx Tx) error) error {
+	return s.db.Update(ctx, f)
+}
+
+func (s *SnapshotKV) Close() {
+	defer s.db.Close()
+	if s.snapshotDB != nil {
+		defer s.snapshotDB.Close()
+	}
+}
+
+func (s *SnapshotKV) Begin(ctx context.Context, parentTx Tx, writable bool) (Tx, error) {
+	dbTx, err := s.db.Begin(ctx, parentTx, writable)
+	if err != nil {
+		return nil, err
+	}
+
+	t := &snapshotTX{
+		dbTX: dbTx,
+		snTX: newVirtualTx(func() (Tx, error) {
+			return s.snapshotDB.Begin(ctx, parentTx, false)
+		}, s.forBuckets),
+		forBuckets: s.forBuckets,
+	}
+	return t, nil
+}
+
+func newVirtualTx(construct func() (Tx, error), forBucket map[string]dbutils.BucketConfigItem) *lazyTx {
+	return &lazyTx{
+		construct:  construct,
+		forBuckets: forBucket,
+	}
+}
+
+var ErrNotSnapshotBucket = errors.New("not snapshot bucket")
+
+//lazyTx is used for lazy transaction creation.
+type lazyTx struct {
+	construct  func() (Tx, error)
+	forBuckets map[string]dbutils.BucketConfigItem
```

Comment:
You seem to be only using it for the key, not for the value. Is it better to convert it to `map[string]struct{}` ?